File: android/app/src/main/kotlin/com/agent/aios/update/UpdateRepositoryImpl.kt
package com.agent.aios.update

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.util.Log
import androidx.core.content.FileProvider
import com.agent.aios.BuildConfig
import com.agent.aios.domain.model.UpdateInfo
import com.agent.aios.domain.model.UpdateResult
import com.agent.aios.domain.repository.UpdateRepository
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class UpdateRepositoryImpl
    @Inject
    constructor(
        @ApplicationContext private val context: Context,
    ) : UpdateRepository {
        private val TAG = "AIOS-UpdateRepo"
        private val api = GitHubReleaseApi(BuildConfig.GITHUB_REPO)

        private val client =
            OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(60, TimeUnit.SECONDS)
                .build()

        override suspend fun checkForUpdate(): UpdateResult {
            return try {
                val release = withContext(Dispatchers.IO) { api.getLatestRelease() }
                val latestVersion = release.tag_name.removePrefix("v")
                val currentVersion = BuildConfig.VERSION_NAME

                val apkAsset =
                    release.assets.find {
                        it.name.endsWith(".apk", ignoreCase = true)
                    } ?: return UpdateResult.NotAvailable

                val isUpdateAvailable = compareVersions(currentVersion, latestVersion) < 0

                UpdateResult.Success(
                    UpdateInfo(
                        isUpdateAvailable = isUpdateAvailable,
                        currentVersion = currentVersion,
                        latestVersion = latestVersion,
                        downloadUrl = apkAsset.browser_download_url,
                        fileSize = apkAsset.size,
                        releaseNotes = release.body,
                        publishedAt = release.published_at,
                    ),
                )
            } catch (e: Exception) {
                Log.e(TAG, "Failed to check for update: ${e.message}")
                UpdateResult.Error(e.message ?: "Unknown error")
            }
        }

        override suspend fun downloadApk(
            url: String,
            fileName: String,
            onProgress: (Float) -> Unit,
        ): File? =
            withContext(Dispatchers.IO) {
                try {
                    val updatesDir = File(context.cacheDir, "updates")
                    if (!updatesDir.exists()) updatesDir.mkdirs()
                    updatesDir.listFiles()?.forEach { it.delete() }

                    val targetFile = File(updatesDir, fileName)
                    val request = Request.Builder().url(url).build()

                    val response = client.newCall(request).execute()
                    if (!response.isSuccessful) {
                        Log.e(TAG, "Download failed: ${response.code}")
                        return@withContext null
                    }

                    val body = response.body ?: return@withContext null
                    val contentLength = body.contentLength()

                    body.byteStream().use { input ->
                        targetFile.outputStream().use { output ->
                            val buffer = ByteArray(8192)
                            var totalRead = 0L
                            var lastProgress = -1f

                            while (true) {
                                val read = input.read(buffer)
                                if (read == -1) break
                                output.write(buffer, 0, read)
                                totalRead += read

                                if (contentLength > 0) {
                                    val progress = (totalRead.toFloat() / contentLength)
                                    if (progress - lastProgress > 0.01f) {
                                        lastProgress = progress
                                        onProgress(progress)
                                    }
                                }
                            }
                        }
                    }

                    onProgress(1f)
                    Log.i(TAG, "Download complete: ${targetFile.length()} bytes")
                    targetFile
                } catch (e: Exception) {
                    Log.e(TAG, "Download error: ${e.message}")
                    null
                }
            }

        override fun canInstallApk(): Boolean {
            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.packageManager.canRequestPackageInstalls()
            } else {
                true
            }
        }

        override fun requestInstallPermissionIntent(): Intent {
            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).apply {
                    data = Uri.parse("package:${context.packageName}")
                }
            } else {
                Intent(Settings.ACTION_SETTINGS)
            }
        }

        override fun installApk(apkFile: File): Boolean {
            return try {
                val uri =
                    FileProvider.getUriForFile(
                        context,
                        "${context.packageName}.fileprovider",
                        apkFile,
                    )
                val intent =
                    Intent(Intent.ACTION_VIEW).apply {
                        setDataAndType(uri, "application/vnd.android.package-archive")
                        addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    }
                context.startActivity(intent)
                Log.i(TAG, "Install intent launched for ${apkFile.name}")
                true
            } catch (e: Exception) {
                Log.e(TAG, "Install failed: ${e.message}")
                false
            }
        }

        private fun compareVersions(
            v1: String,
            v2: String,
        ): Int {
            val parts1 = v1.split(".").map { it.toIntOrNull() ?: 0 }
            val parts2 = v2.split(".").map { it.toIntOrNull() ?: 0 }
            val maxLen = maxOf(parts1.size, parts2.size)
            for (i in 0 until maxLen) {
                val p1 = parts1.getOrElse(i) { 0 }
                val p2 = parts2.getOrElse(i) { 0 }
                if (p1 != p2) return p1.compareTo(p2)
            }
            return 0
        }
    }

File: lib/data/repositories/update_repository_impl.dart
import 'dart:io';

import 'package:dio/dio.dart';
import 'package:open_file/open_file.dart';
import 'package:path_provider/path_provider.dart';

import 'package:aios/data/datasources/remote/github_api.dart';
import 'package:aios/domain/entities/update_info.dart';
import 'package:aios/domain/repositories/update_repository.dart';

class UpdateRepositoryImpl implements UpdateRepository {
  final GitHubApi _api;
  final String _currentVersion;
  final Dio _dio;
  final Future<String> Function() _getCachePath;
  final Future<OpenResult> Function(String path) _openFile;

  UpdateRepositoryImpl({
    required GitHubApi api,
    required String currentVersion,
    required Dio dio,
    Future<String> Function()? getCachePath,
    Future<OpenResult> Function(String path)? openFile,
  }) : _api = api,
       _currentVersion = currentVersion,
       _dio = dio,
       _getCachePath = getCachePath ?? _defaultCachePath,
       _openFile = openFile ?? _defaultOpenFile;

  static Future<String> _defaultCachePath() async {
    final dir = await getTemporaryDirectory();
    return dir.path;
  }

  static Future<OpenResult> _defaultOpenFile(String path) {
    return OpenFile.open(path);
  }

  @override
  Future<UpdateResult> checkForUpdate() async {
    try {
      final release = await _api.getLatestRelease();
      final latestVersion = _stripVersionPrefix(release.tagName);
      final currentVersion = _stripVersionPrefix(_currentVersion);

      if (_compareVersions(latestVersion, currentVersion) <= 0) {
        return const UpdateResult.notAvailable();
      }

      final apkAsset = release.assets
          .where((a) => a.name.endsWith('.apk'))
          .firstOrNull;

      if (apkAsset == null) {
        return const UpdateResult.error('No APK asset found in release');
      }

      return UpdateResult.success(
        UpdateInfo(
          currentVersion: _currentVersion,
          latestVersion: latestVersion,
          downloadUrl: apkAsset.browserDownloadUrl,
          fileSize: apkAsset.size,
          releaseNotes: release.body,
          publishedAt: DateTime.parse(release.publishedAt),
        ),
      );
    } catch (e) {
      print('[AIOS-UpdateRepo] ERROR: checkForUpdate failed: $e');
      return UpdateResult.error(e.toString());
    }
  }

  @override
  Future<String?> downloadApk(
    String url,
    String fileName, {
    void Function(double progress)? onProgress,
  }) async {
    try {
      final cachePath = await _getCachePath();
      final savePath = '$cachePath/$fileName';

      await _dio.download(
        url,
        savePath,
        onReceiveProgress: (received, total) {
          if (total > 0 && onProgress != null) {
            onProgress(received / total);
          }
        },
      );

      final file = File(savePath);
      if (await file.exists()) {
        return savePath;
      }
      return null;
    } catch (e) {
      print('[AIOS-UpdateRepo] ERROR: downloadApk failed: $e');
      return null;
    }
  }

  @override
  Future<bool> installApk(String apkPath) async {
    final file = File(apkPath);
    if (!await file.exists()) return false;
    try {
      final result = await _openFile(apkPath);
      print('[AIOS-UpdateRepo] open_file result: ${result.type}');
      return result.type == ResultType.done;
    } catch (e) {
      print('[AIOS-UpdateRepo] ERROR: installApk failed: $e');
      return false;
    }
  }

  String _stripVersionPrefix(String version) {
    if (version.startsWith('v')) {
      return version.substring(1);
    }
    return version;
  }

  int _compareVersions(String a, String b) {
    final partsA = a.split('.').map(int.parse).toList();
    final partsB = b.split('.').map(int.parse).toList();

    for (var i = 0; i < partsA.length || i < partsB.length; i++) {
      final valA = i < partsA.length ? partsA[i] : 0;
      final valB = i < partsB.length ? partsB[i] : 0;
      if (valA != valB) {
        return valA.compareTo(valB);
      }
    }
    return 0;
  }
}
